src/backend/distributed/commands/table.c

```diff
@@ -399,8 +399,8 @@ PostprocessCreateTableStmtPartitionOf(CreateStmt *createStatement, const
  * ATTACH PARTITION OF command.
  */
 List *
-PostprocessAlterTableStmtAttachPartition(AlterTableStmt *alterTableStatement,
-										 const char *queryString)
+PreprocessAlterTableStmtAttachPartition(AlterTableStmt *alterTableStatement,
```

Comment:
I think the major drawback of switching from Pre hook to post hook, the errors are going to be detected on the workers instead of on the parents.

For example, in the following example, the error is detected on the shards where it is detected on the coordinator before the patch:
```SQL
CREATE TABLE parent (
  event_id serial NOT NULL,-- REFERENCES another_table (x),
  event_time timestamptz NOT NULL DEFAULT now())
  PARTITION BY RANGE (event_time);
SELECT create_distributed_table('parent', 'event_id');


CREATE TABLE non_distributed_child_very_very_very_loooooooooooooooooooooooooooooooooong_table_name (event_id int NOT NULL, event_time timestamptz NOT NULL DEFAULT now(), extra_column_that_doesnt_exist_on_parent int);


ALTER TABLE parent ATTACH PARTITION non_distributed_child_very_very_very_loooooooooooooooooooooooooooooooooong_table_name FOR VALUES FROM ('2021-06-01') TO ('2021-07-30');
ERROR:  table "non_distributed_child_very_very_very_looooooooooooooooooooooooo" contains column "extra_column_that_doesnt_exist_on_parent" not found in parent "parent"
DETAIL:  The new partition may contain only the columns present in parent.
```

In general, we prefer to do things in the post-hook for this reason. 

Do you see any check in `ATExecAttachPartition` that we'd rather prefer to do on the coordinator?

Similarly, when we do things in the post hook, we ensure that the proper locks are acquired by Postgres on the tables. Now, they are not acquired at this step.  Do you see any issues with that?

Reply:
> I think the major drawback of switching from Pre hook to post hook, the errors are going to be detected on the workers instead of on the parents.

I believe the only difference are the errors that may occur while distributing the partition. The other errors will be detected in the coordinator as before.

> in the following example, the error is detected on the shards where it is detected on the coordinator before the patch:

This error is detected in the coordinator right after distributing the partition. Similarly, all checks in `ATExecAttachPartition` will occur in the coordinator before sending remote "attach partition" commands for shards.

> Similarly, when we do things in the post hook, we ensure that the proper locks are acquired by Postgres on the tables. Now, they are not acquired at this step. Do you see any issues with that?

Hmm I don't see any obvious issues myself, but would it be safer if I acquired the proper locks inside the `Preprocess` function?

Reply:
> I believe the only difference are the errors that may occur while distributing the partition. The other errors will be detected in the coordinator as before.

> This error is detected in the coordinator right after distributing the partition. Similarly, all checks in > ATExecAttachPartition will occur in the coordinator when executing ALTER TABLE parent ATTACH > PARTITION child locally, which happens before sending remote "attach partition" commands for shards.


I don't think so. See below, the error is detected on the **workers** after this patch. And, all the checks in `ATExecAttachPartition ` happen similarly.
```SQL
create table part ( a int, list varchar(5) ) partition by list (list);
SELECT create_distributed_table('part', 'a');

create table part_1 ( a int, list varchar(5), x int );

ALTER TABLE part ATTACH PARTITION part_1 FOR VALUES IN (1,2);
ERROR:  table "part_1" contains column "x" not found in parent "part"
DETAIL:  The new partition may contain only the columns present in parent.
```


The order is : 1)Pre hook 2) standard process utility 3)Post hook.

Before this patch, we first call `ATExecAttachPartition` on the coordinator for the distributed tables at (2). Now, we first call `ATExecAttachPartition` on the shards at (1)


> Hmm would it be safer if I acquired the proper locks inside the Preprocess function? I believe there might be different behaviours if the parent table is concurrently changed while the partition is being distributed. Before with the post hook the parent table already has proper locks at the step of distributing the attached partition. @onderkalaci


Yes, we need to make sure that we acquire the locks in the same order as Postgres. Otherwise, we can introduce very weird deadlocks. An example for such unexpected deadlocks that occurs in Citus but not in PG: https://github.com/citusdata/citus/pull/5159

Please work with @velioglu on the locking issues and make sure that we acquire the locks in the same order as before.